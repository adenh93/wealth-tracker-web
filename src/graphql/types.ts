export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type CryptoCurrency = {
  __typename?: 'CryptoCurrency';
  id: Scalars['Int'];
  name: Scalars['String'];
  rank: Scalars['Int'];
  symbol: Scalars['String'];
};

export type CryptoCurrencyHolding = {
  __typename?: 'CryptoCurrencyHolding';
  cryptoCurrency?: Maybe<CryptoCurrency>;
  cryptoCurrencyId: Scalars['Int'];
  holdings: Scalars['Float'];
  id: Scalars['Int'];
  netHoldingsValue: Scalars['Float'];
  percentChange24h: Scalars['Float'];
  price: Scalars['Float'];
};

export type CryptoCurrencyHoldingsAddInput = {
  cryptoCurrencyId: Scalars['Int'];
  holdings: Scalars['Float'];
};

export type CryptoCurrencyHoldingsUpdateInput = {
  holdings: Scalars['Float'];
  id: Scalars['Int'];
};

export type Mutation = {
  __typename?: 'Mutation';
  cryptoCurrencyHoldingsAdd?: Maybe<CryptoCurrencyHolding>;
  cryptoCurrencyHoldingsDelete?: Maybe<CryptoCurrencyHolding>;
  cryptoCurrencyHoldingsUpdate?: Maybe<CryptoCurrencyHolding>;
};


export type MutationCryptoCurrencyHoldingsAddArgs = {
  input: CryptoCurrencyHoldingsAddInput;
};


export type MutationCryptoCurrencyHoldingsDeleteArgs = {
  id: Scalars['Int'];
};


export type MutationCryptoCurrencyHoldingsUpdateArgs = {
  input: CryptoCurrencyHoldingsUpdateInput;
};

export type Query = {
  __typename?: 'Query';
  cryptoCurrencies?: Maybe<Array<CryptoCurrency>>;
  cryptoCurrencyHoldings?: Maybe<Array<CryptoCurrencyHolding>>;
};


export type QueryCryptoCurrenciesArgs = {
  query?: Maybe<Scalars['String']>;
};
