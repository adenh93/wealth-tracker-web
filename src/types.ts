export interface Asset {
  id: string
  logoSrc: string
  name: string
  ticker: string
  price: number
  gain24Hr: number
  holdings: number
}
