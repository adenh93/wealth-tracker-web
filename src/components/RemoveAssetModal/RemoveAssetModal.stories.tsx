import { Story } from '@storybook/react'
import { Box } from '@mui/material'
import { Asset } from '../../types'
import RemoveAssetModal, { RemoveAssetModalProps } from './RemoveAssetModal'

const config = {
  title: 'Components/RemoveAssetModal',
  component: RemoveAssetModal,
}

const Template: Story<RemoveAssetModalProps> = (args) => (
  <Box>
    <RemoveAssetModal {...args} />
  </Box>
)

export const Primary = Template.bind({})

const asset: Asset = {
  id: 'test-id',
  logoSrc: 'https://upload.wikimedia.org/wikipedia/commons/4/4e/BitConnect.png',
  name: 'Bitconnect',
  ticker: 'BCC',
  price: 0.003021,
  gain24Hr: -0.9984,
  holdings: 2450000.15,
}

Primary.args = {
  asset,
  open: true,
  handleDelete: () => {},
  handleClose: () => {},
}

export default config
