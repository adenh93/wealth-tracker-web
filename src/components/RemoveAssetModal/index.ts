export { default } from './RemoveAssetModal'
export type { RemoveAssetModalProps } from './RemoveAssetModal'