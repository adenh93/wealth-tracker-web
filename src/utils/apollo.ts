import { ApolloError } from '@apollo/client'

export const extractInputErrorField = <FormType>(
  error: ApolloError
): keyof FormType | null => {
  if (error.graphQLErrors.length)
    return error.graphQLErrors[0].extensions?.field as keyof FormType
  return null
}
