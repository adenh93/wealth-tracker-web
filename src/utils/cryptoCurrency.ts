import { CMC_LOGO_URL } from '../constants'

export const getCMCLogoUrl = (id: number): string => `${CMC_LOGO_URL}/${id}.png`
