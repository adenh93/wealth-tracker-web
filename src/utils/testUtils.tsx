import { ReactElement } from 'react'
import { ThemeProvider } from '@mui/material'
import { render as tlRender, RenderResult } from '@testing-library/react'
import theme from '../theme'

export const render = (component: ReactElement): RenderResult =>
  tlRender(<ThemeProvider theme={theme}>{component}</ThemeProvider>)
