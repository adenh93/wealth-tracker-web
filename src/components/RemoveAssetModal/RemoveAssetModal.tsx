import { ChangeEvent, FC, useState } from 'react'
import { Button, Input, Typography } from '@mui/material'
import Modal from '../ui/Modal'
import { Asset } from '../../types'
import { ButtonContainer } from './RemoveAssetModal.css'

export interface RemoveAssetModalProps {
  asset: Asset
  open: boolean
  handleDelete: () => void
  handleClose: (e: any) => void
}

const RemoveAssetModal: FC<RemoveAssetModalProps> = ({
  asset,
  open,
  handleDelete,
  handleClose,
}) => {
  const [inputState, setInputState] = useState('')
  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) =>
    setInputState(e.currentTarget.value)

  return (
    <Modal open={open} title="Delete Asset" handleClose={handleClose}>
      <Typography variant="body2" sx={{ mb: 2 }}>
        Are you sure you want to remove the following asset from your portfolio?
        This action cannot be reversed:
      </Typography>
      <Typography variant="body2" fontWeight="bold" sx={{ mb: 2 }}>
        {asset.name} ({asset.ticker})
      </Typography>
      <Typography variant="body2" sx={{ mb: 2 }}>
        If you are sure, enter 'delete' to remove this asset.
      </Typography>
      <Input
        value={inputState}
        onChange={handleInputChange}
        sx={{ mb: 4 }}
        data-testid="delete-input"
      />
      <ButtonContainer>
        <Button
          variant="contained"
          color="error"
          disabled={inputState !== 'delete'}
          onClick={handleDelete}
          data-testid="confirm-button"
        >
          Confirm
        </Button>
        <Button
          variant="contained"
          color="secondary"
          onClick={handleClose}
          data-testid="cancel-button"
        >
          Cancel
        </Button>
      </ButtonContainer>
    </Modal>
  )
}

export default RemoveAssetModal
