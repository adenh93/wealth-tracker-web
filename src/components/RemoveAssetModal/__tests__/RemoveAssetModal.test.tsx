import { Simulate } from 'react-dom/test-utils'
import { waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Asset } from '../../../types'
import { render } from '../../../utils/testUtils'
import RemoveAssetModal from '../RemoveAssetModal'

describe('RemoveAssetModal component tests', () => {
  const mockAsset: Asset = {
    id: 'test-id',
    logoSrc:
      'https://upload.wikimedia.org/wikipedia/commons/4/4e/BitConnect.png',
    name: 'Bitconnect',
    ticker: 'BCC',
    price: 0.003021,
    gain24Hr: -0.9984,
    holdings: 2450000.15,
  }

  test('matches snapshot', () => {
    const { asFragment } = render(
      <RemoveAssetModal
        asset={mockAsset}
        open={true}
        handleDelete={jest.fn()}
        handleClose={jest.fn()}
      />
    )

    expect(asFragment).toMatchSnapshot()
  })

  test('renders asset details correctly', () => {
    const { getByText } = render(
      <RemoveAssetModal
        asset={mockAsset}
        open={true}
        handleDelete={jest.fn()}
        handleClose={jest.fn()}
      />
    )

    expect(getByText('Bitconnect (BCC)')).not.toBeNull()
  })

  test('confirm button is disabled if input value is not "delete"', () => {
    const { getByTestId } = render(
      <RemoveAssetModal
        asset={mockAsset}
        open={true}
        handleDelete={jest.fn()}
        handleClose={jest.fn()}
      />
    )

    const input = getByTestId('delete-input')

    userEvent.type(input, 'not-delete')

    const confirmButton = getByTestId('confirm-button') as HTMLButtonElement

    expect(confirmButton.disabled).toBe(true)
  })

  test('confirm button is enabled if input value equals "delete"', async () => {
    const { getByTestId } = render(
      <RemoveAssetModal
        asset={mockAsset}
        open={true}
        handleDelete={jest.fn()}
        handleClose={jest.fn()}
      />
    )

    const input = getByTestId('delete-input')

    userEvent.type(input, 'delete')

    const confirmButton = getByTestId('confirm-button') as HTMLButtonElement

    await waitFor(() => expect(confirmButton.disabled).toBe(false))
  })

  test('confirm button calls handleDelete when clicked', async () => {
    const mockHandleDelete = jest.fn()

    const { getByTestId } = render(
      <RemoveAssetModal
        asset={mockAsset}
        open={true}
        handleDelete={mockHandleDelete}
        handleClose={jest.fn()}
      />
    )

    const input = getByTestId('delete-input')

    userEvent.type(input, 'delete')

    const confirmButton = getByTestId('confirm-button') as HTMLButtonElement

    Simulate.click(confirmButton)

    expect(mockHandleDelete).toHaveBeenCalled()
  })

  test('cancel button calls handleClose when clicked', () => {
    const mockHandleClose = jest.fn()

    const { getByTestId } = render(
      <RemoveAssetModal
        asset={mockAsset}
        open={true}
        handleDelete={jest.fn()}
        handleClose={mockHandleClose}
      />
    )

    const cancelButton = getByTestId('cancel-button')
    Simulate.click(cancelButton)

    expect(mockHandleClose).toHaveBeenCalled()
  })
})
