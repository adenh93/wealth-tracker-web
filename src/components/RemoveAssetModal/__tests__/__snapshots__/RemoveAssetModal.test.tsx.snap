// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveAssetModal component tests matches snapshot 1`] = `[Function]`;
